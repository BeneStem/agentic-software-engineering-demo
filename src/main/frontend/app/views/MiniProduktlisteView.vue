<template>
  <div v-if="produkte.length > 0" class="erk_finden__mini-produktliste-view">
    <div class="erk_finden__mini-produktliste-view-header">
      <UeberschriftDrei
        class="erk_finden__mini-produktliste-view-headline"
        :ist-haupt-ueberschrift="istHauptueberschrift"
        v-if="ueberschrift"
      >
        {{ ueberschrift }}
      </UeberschriftDrei>
      <a
        class="erk_finden__mini-produktliste-view-alle-anzeigen-link"
        :href="alleAnzeigenLinkZiel"
        v-if="alleAnzeigenLinkText && alleAnzeigenLinkZiel"
        @click="sendTrackAlleAnzeigenKlick"
      >
        <span class="erk_finden__mini-produktliste-view-alle-anzeigen-link-text">{{ alleAnzeigenLinkText }}</span>
        <PfeilspitzeImKreis class="erk_finden__mini-produktliste-view-alle-anzeigen-link-pfeil" richtung="rechts" />
      </a>
    </div>
    <div class="erk_finden__mini-produktliste-view-produkte">
      <MiniProduktliste :produkte="produkte" :trackingname="trackingname" :trackingnummer="trackingnummer" />
    </div>
  </div>
</template>

<script setup lang="ts">
import { computed, onServerPrefetch } from 'vue';
import { useStore } from 'vuex';
import MiniProduktliste from '@/main/frontend/app/components/MiniProduktliste.vue';
import UeberschriftDrei from '@blume2000/design-system/dist/komponenten/content/UeberschriftDrei.vue';
import { useRoute } from 'vue-router';
import { trackAlleAnzeigenKlick } from '../util/tracking';
import PfeilspitzeImKreis from '@blume2000/design-system/dist/komponenten/visuelles/icons/allgemein/PfeilspitzeImKreis.vue';
import { QueryParameters } from '@/main/frontend/app/model/QueryParameters';

const store = useStore();
const ueberschrift = computed(() => store.getters['EinstellungenState/getUeberschrift']);
const istHauptueberschrift = computed(() => store.getters['EinstellungenState/getIstHauptueberschrift']);
const trackingname = computed(() => store.getters['EinstellungenState/getTrackingname']);
const trackingnummer = computed(() => store.getters['EinstellungenState/getTrackingnummer']);
const produkte = computed(() => store.getters['ProdukteState/getProdukte']);
const alleAnzeigenLinkText = computed(() => store.getters['EinstellungenState/getAlleAnzeigenLinkText']);
const alleAnzeigenLinkZiel = computed(() => store.getters['EinstellungenState/getAlleAnzeigenLinkZiel']);

const sendTrackAlleAnzeigenKlick = trackAlleAnzeigenKlick;

onServerPrefetch(async () => {
  const route = useRoute();

  const queryParameters = QueryParameters.mapQueryToQueryParameters(route.query);
  //Das setzten der CMS filter im State muss awaited werden, da fetchProdukte im folgenden auf den CMS filter State zugreift
  await store.commit('ProduktfilterState/setCmsFilter', queryParameters, { root: true });

  return Promise.all([
    store.commit('EinstellungenState/setEinstellungenFromQueryparameters', route.query),
    store.dispatch('FeatureState/fetchFeatureTogglesSSR'),
    store.dispatch('ProdukteState/fetchProdukteSSR'),
  ]);
});
</script>

<style lang="scss" scoped>
.erk_finden__mini-produktliste-view {
  margin: 0 auto;
  max-width: 100%;

  &-header {
    display: flex;
    margin-bottom: 24px;
  }

  &-alle-anzeigen-link {
    @include font-Rubik-medium(18, 25);
    color: $color-dark;

    @media (min-width: 568px) {
      margin-left: auto;
      text-align: right;
      text-decoration: none;
    }

    &-text {
      display: none;

      @media (min-width: 568px) {
        display: inline-block;
      }
    }

    .erk_finden__mini-produktliste-view &-pfeil {
      display: inline-block;
      height: 22px;
      margin-left: 8px;
      width: 22px;

      @media (min-width: 568px) {
        position: relative;
        top: 4px;
      }

      svg {
        height: 22px;
        width: 22px;
      }
    }

    &:visited {
      color: $color-dark;
    }
  }
}
</style>
