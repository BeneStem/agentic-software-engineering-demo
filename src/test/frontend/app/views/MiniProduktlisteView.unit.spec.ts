import { shallowMount } from '@vue/test-utils';
import MiniProduktlisteView from '@/main/frontend/app/views/MiniProduktlisteView.vue';
import { createStore } from 'vuex';

describe('MiniProduktlisteView', () => {
  it('keine Ã¼berschrift wenn undefined', () => {
    // Given
    const store = createMockStore(false, false, false, true);
    const wrapper = shallowMount(MiniProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('zeigt ueberschrift wenn defined', () => {
    // Given
    const store = createMockStore(true, false, false, true);
    const wrapper = shallowMount(MiniProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('zeigt alle anzeigen link wenn text und ziel defined', () => {
    // Given
    const store = createMockStore(true, true, true, true);
    const wrapper = shallowMount(MiniProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('kein alle anzeigen link wenn ziel fehlt', () => {
    // Given
    const store = createMockStore(true, true, false, true);
    const wrapper = shallowMount(MiniProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('kein alle anzeigen link wenn text fehlt', () => {
    // Given
    const store = createMockStore(true, false, true, true);
    const wrapper = shallowMount(MiniProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Sendet beim Klick auf den alle anzeigen Link ein Event an GA', () => {
    // Given
    const store = createMockStore(true, true, true, true);
    const wrapper = shallowMount(MiniProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    wrapper.vm.sendTrackAlleAnzeigenKlick = jest.fn();

    // When
    wrapper.find('a').trigger('click');

    // Then
    expect(wrapper.vm.sendTrackAlleAnzeigenKlick).toHaveBeenCalled();
  });
});

function createMockStore(
  hasUeberschrift: boolean,
  hasLinkText: boolean,
  hasLinkZiel: boolean,
  isFeatureToggleActive: boolean,
  trackingname = 'trackingname',
  trackingnummer = 1234567890
) {
  const store = createStore({
    getters: {
      'EinstellungenState/getUeberschrift': (state: any): string | undefined => {
        return hasUeberschrift ? 'Ã¼berschrift' : undefined;
      },
      'ProdukteState/getProdukte': (state: any) => {
        return [{}];
      },
      'EinstellungenState/getAlleAnzeigenLinkText': (state: any): string | undefined => {
        return hasLinkText ? 'linktext' : undefined;
      },
      'EinstellungenState/getAlleAnzeigenLinkZiel': (state: any): string | undefined => {
        return hasLinkZiel ? 'linkziel' : undefined;
      },
      'FeatureState/getFeatureStatus':
        (state: any) =>
        (featureToggleName: string): boolean => {
          return isFeatureToggleActive;
        },
      'EinstellungenState/getTrackingname': (state: any): string => {
        return trackingname;
      },
      'EinstellungenState/getTrackingnummer': (state: any): number => {
        return trackingnummer;
      }
    },
  });
  store.dispatch = jest.fn();
  return store;
}
