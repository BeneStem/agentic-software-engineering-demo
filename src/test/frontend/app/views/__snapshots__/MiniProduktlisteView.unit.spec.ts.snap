// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiniProduktlisteView kein alle anzeigen link wenn text fehlt 1`] = `
<div
  class="erk_finden__mini-produktliste-view"
>
  <div
    class="erk_finden__mini-produktliste-view-header"
  >
    <ueberschrift-drei-stub
      class="erk_finden__mini-produktliste-view-headline"
      isthauptueberschrift="false"
      istsemantischeueberschrift="true"
      theme="hell"
    />
    <!--v-if-->
  </div>
  <div
    class="erk_finden__mini-produktliste-view-produkte"
  >
    <mini-produktliste-stub
      produkte="[object Object]"
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
  </div>
</div>
`;

exports[`MiniProduktlisteView kein alle anzeigen link wenn ziel fehlt 1`] = `
<div
  class="erk_finden__mini-produktliste-view"
>
  <div
    class="erk_finden__mini-produktliste-view-header"
  >
    <ueberschrift-drei-stub
      class="erk_finden__mini-produktliste-view-headline"
      isthauptueberschrift="false"
      istsemantischeueberschrift="true"
      theme="hell"
    />
    <!--v-if-->
  </div>
  <div
    class="erk_finden__mini-produktliste-view-produkte"
  >
    <mini-produktliste-stub
      produkte="[object Object]"
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
  </div>
</div>
`;

exports[`MiniProduktlisteView keine Ã¼berschrift wenn undefined 1`] = `
<div
  class="erk_finden__mini-produktliste-view"
>
  <div
    class="erk_finden__mini-produktliste-view-header"
  >
    <!--v-if-->
    <!--v-if-->
  </div>
  <div
    class="erk_finden__mini-produktliste-view-produkte"
  >
    <mini-produktliste-stub
      produkte="[object Object]"
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
  </div>
</div>
`;

exports[`MiniProduktlisteView zeigt alle anzeigen link wenn text und ziel defined 1`] = `
<div
  class="erk_finden__mini-produktliste-view"
>
  <div
    class="erk_finden__mini-produktliste-view-header"
  >
    <ueberschrift-drei-stub
      class="erk_finden__mini-produktliste-view-headline"
      isthauptueberschrift="false"
      istsemantischeueberschrift="true"
      theme="hell"
    />
    <a
      class="erk_finden__mini-produktliste-view-alle-anzeigen-link"
      href="linkziel"
    >
      <span
        class="erk_finden__mini-produktliste-view-alle-anzeigen-link-text"
      >
        linktext
      </span>
      <pfeilspitze-im-kreis-stub
        class="erk_finden__mini-produktliste-view-alle-anzeigen-link-pfeil"
        istgefuellt="false"
        richtung="rechts"
        theme="hell"
      />
    </a>
  </div>
  <div
    class="erk_finden__mini-produktliste-view-produkte"
  >
    <mini-produktliste-stub
      produkte="[object Object]"
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
  </div>
</div>
`;

exports[`MiniProduktlisteView zeigt ueberschrift wenn defined 1`] = `
<div
  class="erk_finden__mini-produktliste-view"
>
  <div
    class="erk_finden__mini-produktliste-view-header"
  >
    <ueberschrift-drei-stub
      class="erk_finden__mini-produktliste-view-headline"
      isthauptueberschrift="false"
      istsemantischeueberschrift="true"
      theme="hell"
    />
    <!--v-if-->
  </div>
  <div
    class="erk_finden__mini-produktliste-view-produkte"
  >
    <mini-produktliste-stub
      produkte="[object Object]"
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
  </div>
</div>
`;
